packages/host/app/modifiers/code-block.ts

```diff
@@ -0,0 +1,172 @@
+    editor: MonacoSDK.editor.IStandaloneCodeEditor;
+    model: MonacoSDK.editor.ITextModel;
+    language: string | undefined;
+  }[] = [];
+  modify(
+    element: HTMLElement,
+    _positional: [],
+    {
+      codeBlockSelector,
+      languageAttr,
+      monacoSDK,
+      editorDisplayOptions,
+    }: Signature['Args']['Named'],
+  ) {
+    let codeBlocks = element.querySelectorAll(codeBlockSelector);
+    if (!codeBlocks || codeBlocks.length === 0) {
+      return;
+    }
+
+    for (let [index, codeBlockNode] of [...codeBlocks].entries()) {
+      let codeBlock = codeBlockNode as HTMLElement;
+      let id = codeBlock.id;
+      if (!id) {
+        continue;
+      }
+      // note that since the localstorage item was set from runtime-common there
+      // was no ember-window-mock available, so we can't use window mock here
+      let maybeContent = window.localStorage.getItem(id);
+      window.localStorage.removeItem(id);
+      if (maybeContent == null) {
+        continue;
+      }
+      let content = maybeContent;
+      let lines = content.split('\n').length;
+      let language = codeBlock.getAttribute(languageAttr) ?? undefined;
+      let state = this.monacoState[index];
+      if (state) {
+        let { model, language: lastLanguage } = state;
+        if (language && language !== lastLanguage) {
+          monacoSDK.editor.setModelLanguage(model, language);
+        }
+        if (content !== model.getValue()) {
+          model.setValue(content);
+        }
+      } else {
+        // The light theme editor is used for the main editor in code mode,
+        // but we also have a dark themed editor for the preview editor in AI panel.
+        // The latter is themed using a CSS filter as opposed to defining a new monaco theme
+        // because monaco does not support multiple themes on the same page (check the comment in
+        // room-message-command.gts for more details)
+        monacoSDK.editor.defineTheme('boxel-monaco-light-theme', {
+          base: 'vs',
+          inherit: true,
+          rules: [],
+          colors: {
+            'editor.background': '#FFFFFF',
+          },
+        });
+
+        let editorOptions: MonacoEditorOptions = {
+          readOnly: true,
+          value: content,
+          language,
+          scrollBeyondLastLine: false,
+          automaticLayout: true,
+          minimap: {
+            enabled: false,
+          },
+          theme: 'boxel-monaco-light-theme',
+          ...editorDisplayOptions,
+        };
+
+        // Code rendering is inconsistently wrapped without this,
+        // producing spurious visual diffs
+        if (isTesting()) {
+          editorOptions.wordWrap = 'on';
+        }
+        let monacoContainer = document.createElement('div');
+        monacoContainer.setAttribute(
+          'class',
+          'preview-code monaco-container code-block',
+        );
+        monacoContainer.setAttribute('style', `height: ${lines + 4}rem`);
```

Comment:
This is a simple heuristic to get the codeblock to have a variable height based on the content. The monaco div itself does not automaticlaly grow based on the contents within it. (note that there is a min-height/max-height CSS prop on this to make sure this is well behaved)